Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp

```diff
@@ -20,150 +23,151 @@ static constexpr size_t s_iNotifyMaxEntries = 1024 * 16;         // Control the
     }
+    if (!m_handleToFolderMapLock.tryLock(s_handleToFolderMapLockTimeout))
+    {
+        AZ_Error("FileWatcher", false, "Unable to obtain inotify handle lock on thread");
+        return;
+    }
+    m_handleToFolderMap[watchHandle] = cleanPath;
+    m_handleToFolderMapLock.unlock();
+
+    // Add all the contents (files and directories) to watch and track them
+    QDirIterator dirIter(folder, QDir::NoDotAndDotDot, (recursive ? QDirIterator::Subdirectories : QDirIterator::NoIteratorFlags) | QDirIterator::FollowSymlinks);
 
-    void RemoveWatchFolder(int watchHandle)
+    while (dirIter.hasNext())
     {
-        if (m_iNotifyHandle >= 0)
-        {
-            if (!m_handleToFolderMapLock.tryLock(s_handleToFolderMapLockTimeout))
-            {
-                AZ_Error("FileWatcher", false, "Unable to obtain inotify handle lock on thread");
-                return;
-            }
+        QString dirName = dirIter.next();
 
-            QHash<int, QString>::iterator handleToRemove = m_handleToFolderMap.find(watchHandle);
-            if (handleToRemove != m_handleToFolderMap.end())
-            {
-                inotify_rm_watch(m_iNotifyHandle, watchHandle);
-                m_handleToFolderMap.erase(handleToRemove);
-            }
+        watchHandle = inotify_add_watch(m_iNotifyHandle,
+                                            dirName.toUtf8().constData(),
+                                            IN_CREATE | IN_CLOSE_WRITE | IN_DELETE | IN_DELETE_SELF | IN_MODIFY | IN_MOVE);
+        if (watchHandle < 0)
+        {
+            AZStd::fixed_string<255> errorBuffer;
+            errorBuffer.resize_no_construct(errorBuffer.capacity());
+            const char* errorString = strerror_r(errno, errorBuffer.data(), errorBuffer.size());
+            AZ_Warning("FileWatcher", false, "inotify_add_watch failed for path %s: %s", dirName.toUtf8().constData(), errorString);
+            return;
+        }
 
-            m_handleToFolderMapLock.unlock();
+        if (!m_handleToFolderMapLock.tryLock(s_handleToFolderMapLockTimeout))
+        {
+            AZ_Warning("FileWatcher", false, "Unable to obtain inotify handle lock on thread");
+            return;
         }
+        m_handleToFolderMap[watchHandle] = dirName;
+        m_handleToFolderMapLock.unlock();
     }
-};
-
-//////////////////////////////////////////////////////////////////////////////
-/// FolderWatchRoot
-FolderRootWatch::FolderRootWatch(const QString rootFolder)
-    : m_root(rootFolder)
-    , m_shutdownThreadSignal(false)
-    , m_fileWatcher(nullptr)
-    , m_platformImpl(new PlatformImplementation())
-{
 }
 
-FolderRootWatch::~FolderRootWatch()
+void FileWatcher::PlatformImplementation::RemoveWatchFolder(int watchHandle)
 {
-    // Destructor is required in here since this file contains the definition of struct PlatformImplementation
-    Stop();
+    if (m_iNotifyHandle >= 0)
+    {
+        if (!m_handleToFolderMapLock.tryLock(s_handleToFolderMapLockTimeout))
```

Comment:
I've seen this pattern a fair bit now and again know that it's from old code, but I generally don't like seeing speculative locks where we silently throw away the operation (yes we log but effectively the result is ignored by the caller). What would actually cause us to fail to take this lock?

Reply:
Agreed. 
I would much rather deadlock here, than to silently propagate an issue where the watch folder couldn't be removed because it timed out taking a lock.

It is easier to debug a deadlock than it is to determine after the fact why the lock wasn't taking

Reply:
I've updated the relevant sections to use `QMutexLocker` instead of explicit `lock`/`unlock` calls.